### PR TITLE
Support for separately managing identity provider secrets

### PR DESCRIPTION
**Pull request by ccremer**
This PR is based on #28 

* Adds parameter `.secrets` that expects `stringData`-compatible keys and values.
* Deprecates `ldap.bindPassword` as string. (see also #33)

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues.

### CODE REVIEW

File: tests/golden/defaults/openshift4-authentication/openshift4-authentication/10_oauth.yaml [hunk 1]
```diff
@@ -5,4 +5,31 @@ metadata:
...
+            - email
+          name:
+            - name
+          preferredUsername:
+            - preferred_username
+        clientID: company-client
+        clientSecret:
+          name:
+            clientSecret: t-silent-test-1234/c-green-test-1234/keycloak-auth/clientSecret
+        clientSecretRef: oidc-client
```
Comment (ccremer) [inline tests/golden/defaults/openshift4-authentication/openshift4-authentication/10_oauth.yaml:33]: This parameter isn't needed in the final output. Is this a problem? Can this key be removed? Are there alternatives?

**Review by ccremer — COMMENTED** [review]
I've attempted to do something similar like https://github.com/projectsyn/component-cert-manager/pull/36, but without breaking changes for LDAP provider.

The upgrade is breaking only if OIDC providers are configured.

**Comment by ccremer** [comment]
Technically, OIDC providers are not directly supported by the component (there's no usage in code or docs). Is this even a breaking change or just a new feature?

File: component/main.jsonnet [hunk 2]
```diff
@@ -69,16 +49,29 @@ local identityProviders = [
...
     },
   }
   for idpname in std.objectFields(idps)
   if idps[idpname].type == 'LDAP'
+] + [
+  local idp = idps[idpname];
+  idp {
+    openID+: {
+      clientSecret: {
+        name: params.secrets[idp.openID.clientSecretRef],
```
Comment (simu) [inline component/main.jsonnet:65]:
```suggestion
        name: idp.openID.clientSecretRef,
```

File: component/main.jsonnet [hunk 3]
```diff
@@ -69,16 +49,29 @@ local identityProviders = [
 
   idp {
     ldap+: {
-      ca: { name: authentication.RefName(idp.name) },
-      bindPassword: { name: authentication.RefName(idp.name) },
+      ca: { name: common.RefName(idp.name) },
+      bindPassword: {
+        name: if std.objectHas(idp, 'bindPasswordSecretRef') then params.secrets[idp.bindPasswordSecretRef] else common.RefName(idp.name),
+      },
```
Comment (simu) [inline component/main.jsonnet:55]: Alternative approach to make this backwards compatible without introducing an extra field which might break OCP:

```suggestion
      bindPassword:
        if std.isString(super.bindPassword) then
          // Legacy variant: value of `bindPassword` is a string, so we inject the secret name for the generated legacy secret
          { name: common.RefName(idp.name) }
        else
          // In other cases: Just reuse the original value for `bindPassword`, leave it to the user to correctly format it.
          super.bindPassword,
```

File: component/main.jsonnet [hunk 4]
```diff
@@ -69,16 +49,29 @@ local identityProviders = [
...
+  local idp = idps[idpname];
+  idp {
+    openID+: {
+      clientSecret: {
+        name: params.secrets[idp.openID.clientSecretRef],
+      },
+    },
+  }
+  for idpname in std.objectFields(idps)
+  if idps[idpname].type == 'OpenID'
```
Comment (simu) [inline component/main.jsonnet:70]: This is not necessary, and actually not wanted as the original idea behind Christian's suggestion was that we don't have to explicitly handle each IDP provider in the component (which the approach with `clientSecretRef` actually doesn't achieve), but provide a mechanism which automatically works with all IDP providers.

Just let the users configure `openID.clientSecret.name` to the name of a secret they're configuring with the secrets mechanism.

File: component/ldap.libsonnet [hunk 5]
```diff
@@ -1,13 +1,13 @@
+local common = import 'common.libjsonnet';
```
Comment (simu) [inline component/ldap.libsonnet:1]: In this file, line 23, you need to do something like

```
    bindPassword:
      if std.isString(idp.ldap.bindPassword) then
        idp.ldap.bindPassword
      else
        params.secrets[idp.ldap.bindPassword.name].bindPassword,
```

assuming you go with my suggestion which doesn't need `bindPasswordSecretRef`.

File: component/secrets.jsonnet [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        'argocd.argoproj.io/compare-options': 'IgnoreExtraneous',
+      },
+      labels+: common.commonLabels(),
+    },
+    stringData+: {
+      bindPassword: idp.ldap.bindPassword,
+    },
+  })
+  for idpname in std.objectFields(idps)
+  if idps[idpname].type == 'LDAP' && !std.objectHas(idps[idpname].ldap, 'bindPasswordSecretRef')
```
Comment (simu) [inline component/secrets.jsonnet:28]:
```suggestion
  if idps[idpname].type == 'LDAP' && std.isString(idps[idpname].ldap.bindPassword)
```

**Review by simu — CHANGES_REQUESTED** [review]
Re whether it's a breaking change: potentially not with the adjustments for LDAP to still support the old parameter format, and the fact that we want to support other IDPs without having any code treating them explicitly.

Otherwise, after a quick test, there were a few bugs in the implementation (both LDAP and OpenID), but they are not relevant, if my suggestion of implementing the backwards-compatibility for LDAP is done with the check whether `bindPassword` is a string or not.

File: component/secrets.jsonnet [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+// To avoid breaking changes for LDAP, we support the identityProviders.<name>.ldap.bindPassword parameter
+local legacySecrets = [
+  local idp = idps[idpname];
+
+  com.namespaced(params.namespace, kube.Secret(common.RefName(idp.name)) {
+    metadata+: {
+      annotations+: {
+        'argocd.argoproj.io/compare-options': 'IgnoreExtraneous',
+      },
```
[the remark this answers is not in the document]
Comment (ccremer) [reply component/secrets.jsonnet]: I wasn't sure if `'argocd.argoproj.io/sync-options': 'Prune=false',` is also required.

File: component/main.jsonnet (hunk 4 above)
Comment (ccremer) [reply component/main.jsonnet:70]: Yeah, I got the same idea afterwards. It was friday evening ;)

File: component/secrets.jsonnet [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    },
+  })
+  for idpname in std.objectFields(idps)
+  if idps[idpname].type == 'LDAP' && !std.objectHas(idps[idpname].ldap, 'bindPasswordSecretRef')
+];
+
+local secrets = [
+  com.namespaced(params.namespace, kube.Secret(common.RefName(secretName)) {
+    metadata+: {
+      annotations+: common.argoAnnotations(),
```
[the remark this answers is not in the document]
Comment (ccremer) [reply component/secrets.jsonnet]: thanks

**Comment by ccremer** [comment]
@simu I integrated your feedback and idea, thanks!

So this PR wouldn't be breaking anymore. I'm not quite sure what to do with the documentation. should that simply become a "deprecation" page for the LDAP bindPassword parameter and how to migrate? Or generally create how-to pages for LDAP (legacy + new style syntax) and one for OIDC?

File: component/main.jsonnet (hunk 2 above)
Comment (ccremer) [reply component/main.jsonnet]: With the latest syntax how to configured OIDC this special case basically goes away.

**Comment by simu** [comment]
For this PR, I'd suggest to create describing how to configure an IDP using the secrets mechanism (link to the OCP docs regarding required keys in the credentials secrets), and add a page noting the deprecation of the legacy syntax for LDAP including the snippet on how to transform a legacy LDAP config with a config that works with the new format.

As a follow up, we should add parameters documentation, which is currently missing. There we can then refer to the relevant how-tos for configuring IDPs and how to use secrets.

File: component/main.jsonnet (hunk 2 above)
Comment (simu) [reply component/main.jsonnet]: yep, I added this comment before suggesting my change (this is one of the bugs that I mentioned in my review).

**Comment by ccremer** [comment]
I was re-reading https://syn.tools/commodore/reference/deprecation-policy.html.

> In addition, Commodore will print warnings ("deprecation notices") to the standard output for all deprecated features which are used by a cluster configuration.

What is the way to do that? I'm aware of https://syn.tools/commodore/reference/component-deprecation.html but that's not quite the same.

**Comment by ccremer** [comment]
> add a page noting the deprecation of the legacy syntax for LDAP including the snippet on how to transform a legacy LDAP

I made #33 to have a dedicated changelog entry.

**Comment by simu** [comment]
> I was re-reading https://syn.tools/commodore/reference/deprecation-policy.html.
> 
> > In addition, Commodore will print warnings ("deprecation notices") to the standard output for all deprecated features which are used by a cluster configuration.
> 
> What is the way to do that? I'm aware of https://syn.tools/commodore/reference/component-deprecation.html but that's not quite the same.

There's currently no support to do this for component features, only Commodore features

File: docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+This guide describes the steps to perform an upgrade of the component from version 2.x to 3.x.
+
+Version 3.0 changes how secrets are managed and configured for the LDAP IdP.
+
+
+== Upgrade guide for LDAP provider
+
+With version 3.x, the parameter `openshift4_authentication.identityProviders.<name_of_the_provider>.ldap.bindPassword` as `string` was marked **deprecated** and has been removed with this release.
+
+See xref:references/deprecation-notice.adoc[Deprecated features] how to upgrade.
```
Comment (ccremer) [inline docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc:16]: Note: This page doesn't exist in this PR (see #33)

Comment (simu) [reply docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc:16]: This page is not required anymore, looking at #33, as you add the deprecation notice for v2.3.0 and not v3.0.0.

File: docs/modules/ROOT/pages/how-tos/configure-oidc-provider.adoc [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            name:
+              - name
+            email:
+              - email
+
+    secrets:
+      company-keycloak: <1>
+        clientSecret: '?{vaultkv:${cluster:tenant}/${cluster:name}/oidc/keycloak-auth/clientSecret}'
+----
+<1> Those names have to match
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/configure-oidc-provider.adoc:39]:
```suggestion
<1> The name of the secret.
The name must match between the identity provider configuration and the secret configuration.
<2> For OpenID connect, the client secret **must** be stored in a key named `clientSecret` in the secret.
```

File: docs/modules/ROOT/pages/how-tos/configure-oidc-provider.adoc [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+            preferredUsername:
+              - preferred_username
+            name:
+              - name
+            email:
+              - email
+
+    secrets:
+      company-keycloak: <1>
+        clientSecret: '?{vaultkv:${cluster:tenant}/${cluster:name}/oidc/keycloak-auth/clientSecret}'
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/configure-oidc-provider.adoc:37]:
```suggestion
        clientSecret: '?{vaultkv:${cluster:tenant}/${cluster:name}/oidc/keycloak-auth/clientSecret}' <2>
```

File: tests/defaults.yml [hunk 12]
```diff
@@ -12,3 +12,41 @@ parameters:
...
+      ldap-auth:
+        name: Company LDAP
+        type: LDAP
+        ldap:
+          url: "ldaps://ldap.company.tld:636/ou=services,dc=company,dc=tld?uid"
+          bindDN: "uid=service,ou=idp,dc=company,dc=tld"
+          ca: |-
+            -----BEGIN CERTIFICATE-----
+            -----END CERTIFICATE-----
+          sync: {}
```
Comment (simu) [inline tests/defaults.yml:33]: Just a thought: Might be worth adding a simple LDAP sync config in the default test to verify that the sync config is compiled correctly as well.

**Review by simu — APPROVED** [review]
LGTM now, some comments on the documentation, and from what I can see the "upgrade v2 to v3" documentation page is not required anymore.

File: docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc (hunk 9 above)
Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc:16]: Did you see the note at the top of this page? I left it purposefully there as preparation so it doesn't get forgotten on the next major release.

File: docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc [hunk 13]
```diff
@@ -0,0 +1,16 @@
+// NOTE: this page is a preparation for the next major release by removing deprecated features.
+// It shall only be linked in the nav once a major release is pending.
+
+= Upgrade from v2 to v3
+
+
+This guide describes the steps to perform an upgrade of the component from version 2.x to 3.x.
+
+Version 3.0 changes how secrets are managed and configured for the LDAP IdP.
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc:9]:
```suggestion
Version 3.0 of the component removes the legacy syntax for configuring the bind password for LDAP identity providers.
```

File: docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+
+== Upgrade guide for LDAP provider
+
+With version 3.x, the parameter `openshift4_authentication.identityProviders.<name_of_the_provider>.ldap.bindPassword` as `string` was marked **deprecated** and has been removed with this release.
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/upgrade-v2-to-v3.adoc:14]:
```suggestion
Starting with version 2.3 of the component, the parameter `openshift4_authentication.identityProviders.<name_of_the_provider>.ldap.bindPassword` as `string` was marked **deprecated** and has been removed with this release.
```

**Review by simu — APPROVED** [review]

**Pull request merged**